### PR TITLE
LoupeTarget with support for include MessageDetails in Json-format

### PR DESCRIPTION
**Pull request by snakefoot**
Added `IncludeEventProperties` + `IncludeMldc` + `ContextProperties`.

Possible to do this:

```xml
<target name="Loupe" xsi:type="Loupe" includeEventProperties="true">
    <contextProperty name="threadid" layout="${threadid}" />
</target>
```

This will match `Loupe.Serilog` way of resolving details by including LogEvent-Properties.

### CODE REVIEW

**Comment by snakefoot** [comment]
Example of documentation update:

Adjust your NLog config to include NLog target for Loupe
----------------------------------------------------

Loupe is intended as a top-level catch-all to collect all of your logging in one managed place so you can filter
as needed dynamically during analysis.  To do this, modify the _extensions_, _targets_, and
_rules_ sections of your NLog configuration as follows:

**Loupe.Agent.NLog**
Latest and recommended version with support for .NET Core and improved structured logging capabilities.

1. Install nuget-package:
```
Install-Package Loupe.Agent.NLog
```

2. Update NLog config:
```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <extensions>
        <!--Register nuget-package assembly for LoupeTarget class with target type "Loupe" -->
        <add assembly="Loupe.Agent.NLog" />
    </extensions>
    <targets>
        <!--Define a named target using the "Loupe" target type-->
        <target name="Loupe" xsi:type="Loupe" />
    </targets>
    <rules>
        <!--Send all logging to the "Loupe" named target-->
        <logger name="*" minlevel="Trace" writeTo="Loupe" />
    </rules>
</nlog>
```

3. Start logging
```
    NLog.LogManager.GetLogger("App").Info("Starting...");
```

4. Additional options and Structured logging
NLog 4.5 includes additional support for [structured logging](https://github.com/NLog/NLog/wiki/How-to-use-structured-logging):

_Loupe.Agent.NLog_ has the following options to include additional message details:

- _IncludeEventProperties_ - Include LogEvent properties. Default: _false_
- _IncludeMdlc_ - Include NLog Mapped Diagnostic Logical Context (MDLC) properties. Default: _false_
- _IncludeCallSite_ - Capture and include NLog CallSite information. Default: _true_
- _ContextProperties_ - Capture additional properties with help from [NLog Layout-Renderers](https://nlog-project.org/config/?tab=layout-renderers).

Example of using these options:

```xml
    <target name="Loupe" xsi:type="Loupe" includeEventProperties="true" includeMdlc="true">
        <contextProperty name="threadid" layout="${threadid}" />    <!-- Additional context properties -->
    </target>
```

**Gibraltar.Agent.NLog4**
Support NLog 4.4 (and older) along with .NET Framework ver. 4.5

1. Install nuget-package:
```
Install-Package Gibraltar.Agent.NLog4
```

2. Update NLog config:
```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <extensions>
        <!--Register nuget-package assembly for LoupeTarget class with target type "Loupe" -->
        <add assembly="Gibraltar.Agent.NLog4" />
    </extensions>
    <targets>
        <!--Define a named target using the "Loupe" target type-->
        <target name="Loupe" xsi:type="Loupe" />
    </targets>
    <rules>
        <!--Send all logging to the "Loupe" named target-->
        <logger name="*" minlevel="Trace" writeTo="Loupe" />
    </rules>
</nlog>
```

3. Start logging
```
    NLog.LogManager.GetLogger("App").Info("Starting...");
```

**Gibraltar.Agent.NLog2**
Legacy support for NLog ver. 2 and 3

1. Install nuget-package:
```
Install-Package Gibraltar.Agent.NLog2
```

2. Update NLog config:
```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <extensions>
        <!--Register nuget-package assembly for LoupeTarget class with target type "Loupe" -->
        <add assembly="Gibraltar.Agent.NLog2" />
    </extensions>
    <targets>
        <!--Define a named target using the "Loupe" target type-->
        <target name="Loupe" xsi:type="Loupe" />
    </targets>
    <rules>
        <!--Send all logging to the "Loupe" named target-->
        <logger name="*" minlevel="Trace" writeTo="Loupe" />
    </rules>
</nlog>
```

3. Start logging
```
    NLog.LogManager.GetLogger("App").Info("Starting...");
```

Best Practises
------------------
We recommend to have the NLog logging-rule for the Loupe-target as the first rule,
or at least before any "final" rules - to avoid missing any relevant logevents.

If you find minlevel="Trace" too voluminous for normal collection, the minlevel may be set at
"Debug" instead.  Also, see the example App.config file in the Gibraltar.Agent.NLog4 project,
and consult the [NLog configuration documentation](http://nlog-project.org) as needed.

Make sure to log something to Loupe early in your application to load the Loupe
Agent and initialize its automatic monitoring features.  For example:</p>

```C#
mainLogger.Info("Entering application.");
```

File: src/Agent.NLog45/LoupeTarget.cs [hunk 1]
```diff
@@ -123,77 +221,135 @@ private static LogMessageSeverity GetSeverityLevel(LogLevel level)
...
             /// Should return the line within the file at which the log message was issued.
             /// </summary>
-            public int LineNumber { get; }
+            public int LineNumber { get; private set; }
+        }
+
+        /// <summary>
+        /// IList-Proxy for ContextProperties to skip allocation and rendering using JsonLayout unless requested
+        /// </summary>
+        private class ContextPropetiesProxy : IList<JsonAttribute>, IList
```
Comment (kendallmiller) [inline src/Agent.NLog45/LoupeTarget.cs:280]: ContextPropertiesProxy (missing "r")

Comment (snakefoot) [reply src/Agent.NLog45/LoupeTarget.cs]: Fixed.

**Comment by kendallmiller** [comment]
That doc update looks good but I'm not seeing it in the PR or a different PR; care to put in a PR for it?

**Comment by snakefoot** [comment]
Created PR #14

**Pull request merged**